contracts/upgradeable_contracts/factories/ForeignBridgeFactory.sol

```diff
@@ -0,0 +1,189 @@
+    proxy.transferProxyOwnership(foreignBridgeProxyOwner());
+    // emit event
+    emit ForeignBridgeDeployed(foreignBridge, bridgeValidators, block.number);
+  }
+  
+
+  function bridgeValidatorsImplementation() public view returns(address) {
+    return addressStorage[keccak256(abi.encodePacked("bridgeValidatorsImplementation"))];
+  }
+
+  function setBridgeValidatorsImplementation(address _bridgeValidatorsImplementation) public onlyOwner {
+    addressStorage[keccak256(abi.encodePacked("bridgeValidatorsImplementation"))] = _bridgeValidatorsImplementation;
+  }
+
+  function requiredSignatures() public view returns(uint256) {
+    return uintStorage[keccak256(abi.encodePacked("requiredSignatures"))];
+  }
+
+  function setRequiredSignatures(uint256 _requiredSignatures) public onlyOwner {
+    uintStorage[keccak256(abi.encodePacked("requiredSignatures"))] = _requiredSignatures;
+  }
+
+  function initialValidators() public view returns(address[]) {
+    return addressArrayStorage[keccak256(abi.encodePacked("initialValidators"))];
+  }
+
+  function setInitialValidators(address[] _initialValidators) public onlyOwner {
+    addressArrayStorage[keccak256(abi.encodePacked("initialValidators"))] = _initialValidators;
+  }
+
+  function bridgeValidatorsOwner() public view returns(address) {
+    return addressStorage[keccak256(abi.encodePacked("bridgeValidatorsOwner"))];
+  }
+
+  function setBridgeValidatorsOwner(address _bridgeValidatorsOwner) public onlyOwner {
+    addressStorage[keccak256(abi.encodePacked("bridgeValidatorsOwner"))] = _bridgeValidatorsOwner;
+  }
+
+  function bridgeValidatorsProxyOwner() public view returns(address) {
+    return addressStorage[keccak256(abi.encodePacked("bridgeValidatorsProxyOwner"))];
+  }
+
+  function setBridgeValidatorsProxyOwner(address _bridgeValidatorsProxyOwner) public onlyOwner {
+    addressStorage[keccak256(abi.encodePacked("bridgeValidatorsProxyOwner"))] = _bridgeValidatorsProxyOwner;
+  }
+
+  function foreignBridgeErcToErcImplementation() public view returns(address) {
+    return addressStorage[keccak256(abi.encodePacked("foreignBridgeErcToErcImplementation"))];
+  }
+
+  function setForeignBridgeErcToErcImplementation(address _foreignBridgeErcToErcImplementation) public onlyOwner {
+    addressStorage[keccak256(abi.encodePacked("foreignBridgeErcToErcImplementation"))] = _foreignBridgeErcToErcImplementation;
+  }
+
+  function requiredBlockConfirmations() public view returns(uint256) {
+    return uintStorage[keccak256(abi.encodePacked("requiredBlockConfirmations"))];
+  }
+
+  function setRequiredBlockConfirmations(uint256 _requiredBlockConfirmations) public onlyOwner {
+    uintStorage[keccak256(abi.encodePacked("requiredBlockConfirmations"))] = _requiredBlockConfirmations;
+  }
+
+  function gasPrice() public view returns(uint256) {
+    return uintStorage[keccak256(abi.encodePacked("gasPrice"))];
+  }
+
+  function setGasPrice(uint256 _gasPrice) public onlyOwner {
```

Comment:
It looks like there are some methods that are repeated between `HomeBridgeFactory` and `ForeignBridgeFactory`. Does it makes sense to define a common contract where this shared logic could be stored and then extend from that contract? You can take a look at `BasicBridge` as an example https://github.com/poanetwork/poa-bridge-contracts/blob/master/contracts/upgradeable_contracts/BasicBridge.sol

Reply:
good idea, will do